<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Macro to define a D435 Gazebo sensor -->
  <xacro:macro name="camera_gazebo_sensor" params="parent">
    <!-- Fixed joint connecting camera_link to parent -->
    <joint name="camera_joint" type="fixed">
      <parent link="${parent}"/>
      <child link="camera_link"/>
      <origin xyz="0.090925 0.0 0.06670" rpy="0 0 0" />
    </joint>
    
    <link name="camera_link">
      <!-- Simple visual representation of the camera -->
      <visual>
        <geometry>
          <box size="0.01 0.01 0.01"/>
        </geometry>
        <material name="Red"/>
      </visual>
    </link>

    <joint name="camera_optical_joint" type="fixed">
        <parent link="camera_link"/>
        <child link="camera_link_optical"/>
        <origin xyz="0 0 0" rpy="${-M_PI_2} 0 ${-M_PI_2}"/>
    </joint>

    <link name="camera_link_optical"></link>

    <!-- Gazebo simulation properties -->
    <gazebo reference="camera_link">

        <material>Gazebo/Red</material>

        <sensor name="camera" type="camera">
            <always_on>true</always_on>
            <visualize>true</visualize>
            <update_rate>10</update_rate>
            <topic>camera/image_raw</topic>
            <camera>
                <horizontal_fov>1.089</horizontal_fov>
                <image>
                    <format>R8G8B8</format>
                    <width>640</width>
                    <height>480</height>
                </image>
                <clip>
                    <near>0.05</near>
                    <far>8.0</far>
                </clip>
            </camera>
        </sensor>
    </gazebo>

  </xacro:macro>

</robot>
